### PR TITLE
Enabled LRFinder to run multiple epochs

### LINKED ISSUE #1357: FastaiLRFinder does not run more than 1 epoch. Why?
I have been trying to use the `FastaiLRFinder` to find the best learning rate for my module.

If we create the trainer with the function `create_supervised_trainer` as below:

```
trainer = create_supervised_trainer(
    model, optimizer, criterion, device, output_transform=custom_output_transform
)
```

and run it:

```
with lr_finder.attach(
            trainer,
            to_save=to_save,
            num_iter=50,  
            end_lr=1.,
            step_mode='exp') as lr_finder_training:
        lr_finder_training.run(train_loader)
```

A warning will come up say: "UserWarning: Desired num_iter 50 is unreachable with the current run setup of 15 iteration (1 epochs) from ignite.contrib.handlers.param_scheduler import (LRScheduler, PiecewiseLinear)"

My dataloader has 15 batches to iterate, which means FastaiLRFinder does not allow you to run more than 1 epoch. Why?

According to their source code, we can see [here](https://github.com/pytorch/ignite/blob/14cdc046c7cac2db8a1bf838dcd09cbbe0e9c33b/ignite/contrib/handlers/lr_finder.py#L87) that this verification limits the user to run more iterations than the maximum of epochs in their dataloader.

But why? Am I missing something important here?

### PR DESCRIPTION
**Pull request by KickItLikeShika**
Fixes #1357

Previously LRFinder raises a warning if `num_iter` was equivalent to more than 1 epoch, fixed that, and it runs more than 1 epoch now.

Check list:

- [x] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/handlers/test_lr_finder.py [hunk 1]
```diff
@@ -321,6 +321,18 @@ def test_detach_terminates(lr_finder, to_save, dummy_engine, dataloader, recwarn
...
+def test_different_num_iters(lr_finder, to_save, dummy_engine, dataloader):
+    with lr_finder.attach(dummy_engine, to_save, num_iter=200) as trainer_with_finder:
+        trainer_with_finder.run(dataloader)
+    len_lrs_few_iters = len(lr_finder._history["lr"])
+
+    with lr_finder.attach(dummy_engine, to_save, num_iter=1000) as trainer_with_finder:
+        trainer_with_finder.run(dataloader)
+    len_lrs_many_iters = len(lr_finder._history["lr"])
+
+    assert len_lrs_many_iters - len_lrs_few_iters > 500
```
Comment (vfdev-5) [inline tests/ignite/handlers/test_lr_finder.py:333]: I think the test should ensure that trainer_with_finder.state.iteration after the run is equal to num_iter

File: ignite/handlers/lr_finder.py [hunk 2]
```diff
@@ -103,12 +103,10 @@ def _run(
...
-            if num_iter > max_iter:
-                warnings.warn(
-                    f"Desired num_iter {num_iter} is unreachable with the current run setup of {max_iter} iteration "
-                    f"({trainer.state.max_epochs} epochs)",
-                    UserWarning,
-                )
+            if max_iter < num_iter:
+                max_iter = num_iter
+                trainer.state.max_iters = num_iter
+                trainer.state.max_epochs = num_iter // trainer.state.epoch_length  # type: ignore[operator]
```
Comment (vfdev-5) [inline ignite/handlers/lr_finder.py:109]: I wonder if this should not be ceil instead of floor...

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @KickItLikeShika 
I left few comments to improve the PR

File: tests/ignite/handlers/test_lr_finder.py (hunk 1 above)
Comment (KickItLikeShika) [reply tests/ignite/handlers/test_lr_finder.py:333]: I tried that, the number of iterations always is not 100% equal to `num_iters`, it's always slightly smaller (that was always the case, even before these updates), how much smaller depends on the scheduler we are using, so If we would like to this, we should do range, for example
```py
num_iter = 100
num_iter - trainer_with_finder.state.iteration < 10
```

File: ignite/handlers/lr_finder.py (hunk 2 above)
Comment (KickItLikeShika) [reply ignite/handlers/lr_finder.py:109]: I think it should stay floor, for example 150 num_iters should be 1 epoch I think, and have a look here here for example https://github.com/pytorch/ignite/blob/master/tests/ignite/handlers/test_lr_finder.py#L306

File: tests/ignite/handlers/test_lr_finder.py (hunk 1 above)
Comment (vfdev-5) [reply tests/ignite/handlers/test_lr_finder.py:333]: That's true ! Maybe we can set the divergence threshold rather high such that we wont exit before ? In this case they should be exact. Or specify end_lr not that high such that loss value wont go too high neither ...

File: ignite/handlers/lr_finder.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/handlers/lr_finder.py:109]: Thanks for the link. I think this test shows that the API is a bit weird:
- dataloader size is 100
- user requires 150 iters to do
```python
    with lr_finder.attach(dummy_engine, to_save, num_iter=150, diverge_th=float("inf")) as trainer_with_finder:
        with pytest.warns(UserWarning):
            trainer_with_finder.run(dataloader)
        # We should do 150 iters and not the size of the dataloader:
        assert dummy_engine.state.iteration != len(dataloader)
        assert dummy_engine.state.iteration == 150
```

Comment (KickItLikeShika) [reply ignite/handlers/lr_finder.py:109]: Thanks for the clarification, you're right, this test should be edited according to the updates, as the lrfinder now is able to run more than one epoch, will edit it

**Review by vfdev-5 — APPROVED** [review]
Thanks for the PR, Ahmed ! Looks good
